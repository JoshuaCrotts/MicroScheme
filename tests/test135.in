;;; 2.2.3

(define (append list1 list2)
    (if (null? list1)
        list2
        (cons (car list1) (append (cdr list1) list2))
    )
)

(define (odd? x) (= 1 (remainder x 2)))
(define (filter predicate sequence)
  (cond ((null? sequence) '())
        ((predicate (car sequence))
         (cons (car sequence)
               (filter predicate (cdr sequence))))
        (else (filter predicate (cdr sequence)))))
(filter odd? (list 1 2 3 4 5))
; expect (1 3 5)

(define (accumulate op initial sequence)
  (if (null? sequence)
      initial
      (op (car sequence)
          (accumulate op initial (cdr sequence)))))
(accumulate (λ (x y) (+ x y)) 0 (list 1 2 3 4 5))
; expect 15

(accumulate (λ (x y) (* x y)) 1 (list 1 2 3 4 5))
; expect 120

(accumulate (λ (x y) (cons x y)) '() (list 1 2 3 4 5))
; expect (1 2 3 4 5)

(define (enumerate-interval low high)
  (if (> low high)
      '()
      (cons low (enumerate-interval (+ low 1) high))))
(enumerate-interval 2 7)
; expect (2 3 4 5 6 7)

(define (enumerate-tree tree)
  (cond ((null? tree) '())
        ((not (pair? tree)) (list tree))
        (else (append (enumerate-tree (car tree))
                      (enumerate-tree (cdr tree))))))
(enumerate-tree (list 1 (list 2 (list 3 4)) 5))
; expect (1 2 3 4 5)
