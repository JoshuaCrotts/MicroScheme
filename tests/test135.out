(1 3 5)
15
120
(1 2 3 4 5)
(2 3 4 5 6 7)
(1 2 3 4 5)
